### PR TITLE
Custom Header and Response Status Check

### PR DESCRIPTION
**Pull request by tatashidayat**
Hi @OutdatedGuy,
I found that this package is awesome, but in my usecase I need custom headers and response callback to determine the connection status.

So I think it would be great to add those to this package.

Please help to review it.

### CODE REVIEW

File: lib/src/internet_check_option.dart [hunk 1]
```diff
@@ -1,5 +1,7 @@
 part of internet_connection_checker_plus;
 
+typedef ResponseStatusFn = bool Function(http.Response response);
+
```
Comment (OutdatedGuy) [inline lib/src/internet_check_option.dart:4]: Please add documentation for this public definition.

File: lib/src/internet_check_option.dart [hunk 2]
```diff
@@ -27,12 +29,17 @@ class InternetCheckOption {
...
+  ///   },
   /// );
   /// ```
   InternetCheckOption({
     required this.uri,
     this.timeout = const Duration(seconds: 3),
-  });
+    Map<String, String>? headers,
+    this.responseStatusFn,
+  }) : headers = headers ?? <String, String>{};
```
Comment (OutdatedGuy) [inline lib/src/internet_check_option.dart:42]: Why initialize headers like this? We can just use:

```dart
this.headers = const {},
```

File: lib/src/internet_check_option.dart [hunk 3]
```diff
@@ -50,11 +57,18 @@ class InternetCheckOption {
   /// Defaults to 3 seconds.
   final Duration timeout;
 
+  /// A map of additional headers to send with the request.
+  final Map<String, String> headers;
+
+  /// A callback to check expected response
+  final ResponseStatusFn? responseStatusFn;
```
Comment (OutdatedGuy) [inline lib/src/internet_check_option.dart:64]: Can you elaborate the documentation here more.

e.g. A custom callback function to decide whether the request succeeded or not.

Something like this 👆🏼

File: lib/src/internet_connection.dart [hunk 4]
```diff
@@ -127,11 +127,19 @@ class InternetConnection {
     InternetCheckOption option,
   ) async {
     try {
-      final response = await http.head(option.uri).timeout(option.timeout);
+      final response = await http
+          .head(
+            option.uri,
+            headers: option.headers,
```
Comment (OutdatedGuy) [inline lib/src/internet_connection.dart:133]: Please remove the trailing comma on this line and run:

```bash
dart format .
```

File: lib/src/internet_connection.dart [hunk 5]
```diff
@@ -127,11 +127,19 @@ class InternetConnection {
...
+          )
+          .timeout(option.timeout);
+
+      ResponseStatusFn statusFn =
+          option.responseStatusFn ?? (response) => response.statusCode == 200;
```
Comment (OutdatedGuy) [inline lib/src/internet_connection.dart:138]: Please change this to:

```dart
defaultResponseStatusFn(response) => response.statusCode == 200;
final statusFn = option.responseStatusFn ?? defaultResponseStatusFn;
```

File: test/__mocks__/test_http_client.dart
Comment (OutdatedGuy) [inline test/__mocks__/test_http_client.dart:1]: Why is this file necessary? I removed the code from `test/internet_connection_test.dart` referencing this file and ran the tests and it works properly based on internet connection.

File: test/internet_connection_test.dart [hunk 6]
```diff
@@ -21,6 +23,55 @@ void main() {
       expect(await checker.hasInternetAccess, false);
     });
 
+    test('hasInternetAccess invoke responseStatusFn to determine success',
+        () async {
+      await TestHttpClient.run((client) async {
+        client.responseBuilder =
+            (req) => TestHttpClient.createResponse(statusCode: 500);
```
Comment (OutdatedGuy) [inline test/internet_connection_test.dart:30]: Why is this code necessary? The tests are running properly without these additional Mock codes.

File: test/internet_connection_test.dart [hunk 7]
```diff
@@ -21,6 +23,55 @@ void main() {
...
+        client.responseBuilder = (req) {
+          for (final header in expectedHeaders.entries) {
+            final key = header.key;
+            if (!req.headers.containsKey(key) ||
+                req.headers[key] != header.value) {
+              return TestHttpClient.createResponse(statusCode: 500);
+            }
+          }
+          return TestHttpClient.createResponse(statusCode: 200);
+        };
```
Comment (OutdatedGuy) [inline test/internet_connection_test.dart:60]: Why is this code necessary? The tests are running properly without these additional Mock codes.

**Review by OutdatedGuy — CHANGES_REQUESTED** [review]
Yo @tatashidayat, thanks for this PR. It looks good on the logic and is working properly.

I just need a few necessary changes mentioned above.

**Comment by OutdatedGuy** [comment]
Hi @tatashidayat, do you plan to follow up on this PR?

File: lib/src/internet_check_option.dart (hunk 1 above)
Comment (tatashidayat) [reply lib/src/internet_check_option.dart:4]: sure will do

File: lib/src/internet_check_option.dart (hunk 2 above)
Comment (tatashidayat) [reply lib/src/internet_check_option.dart:42]: sure, will modify it

File: test/internet_connection_test.dart (hunk 6 above)
Comment (tatashidayat) [reply test/internet_connection_test.dart:30]: This code to make sure we didn't connect to internet. Assuming we have air gapped CI env.

File: test/internet_connection_test.dart (hunk 7 above)
Comment (tatashidayat) [reply test/internet_connection_test.dart:60]: It is use to verify our expected header was exist on the request.

File: lib/src/internet_connection.dart (hunk 4 above)
Comment (tatashidayat) [reply lib/src/internet_connection.dart:133]: Sorry, I was used to use tralling comma. If its not the style, I'll remove it.

**Comment by tatashidayat** [comment]
Hi @OutdatedGuy, I haven't seen this PR for a while.

Let address some comments.

File: lib/src/internet_connection.dart (hunk 4 above)
Comment (tatashidayat) [reply lib/src/internet_connection.dart]: i've reformat it, please review it again

File: lib/src/internet_connection.dart (hunk 5 above)
Comment (tatashidayat) [reply lib/src/internet_connection.dart]: I've assign default responseStatusFn on the constructor, so we don't need tho define it here

File: lib/src/internet_check_option.dart (hunk 3 above)
Comment (tatashidayat) [reply lib/src/internet_check_option.dart]: docs added

File: lib/src/internet_check_option.dart (hunk 1 above)
Comment (OutdatedGuy) [reply lib/src/internet_check_option.dart:5]: You mistakenly used `//` instead of `///` here.

**Review by OutdatedGuy — APPROVED** [review]
LGTM!
